### ISSUE TITLE
Bug in label smoothing with ignored samples

### ISSUE THREAD

**Issue by TomerRonen34** [pytorch/pytorch#73205]
### 🐛 Describe the bug

`CrossEntropyLoss` with `label_smoothing` gives incorrect results if the input has ignored samples: samples that have target=IGNORE_INDEX (usually -100).
This is because the uniform part of the loss incorrectly takes into account ignored samples, unlike the one-hot part of the loss which works fine.

I verified that with `label_smoothing != 0.0` both "mean" and "sum" reductions fail, and that my code is able to reproduce the default (wrong) behavior:
```
label_smoothing = 0.0, reduction = sum, is correct? True, behavior mimicked? True
label_smoothing = 0.0, reduction = mean, is correct? True, behavior mimicked? True
label_smoothing = 0.2, reduction = sum, is correct? False, behavior mimicked? True
label_smoothing = 0.2, reduction = mean, is correct? False, behavior mimicked? True
```

### Code for reproducing
```
from typing import Tuple

import torch
from torch import tensor, randperm, Tensor
from torch.nn import CrossEntropyLoss


def label_smoothing_bug():
    with torch.no_grad():
        log_probs, target = _create_inputs()
        log_probs_masked = log_probs[target != -100]
        target_masked = target[target != -100]

        for label_smoothing in [0.0, 0.2]:
            for reduction in ["sum", "mean"]:
                loss_obj = CrossEntropyLoss(reduction=reduction, label_smoothing=label_smoothing)
                standard_loss = loss_obj(log_probs, target)
                masked_loss = loss_obj(log_probs_masked, target_masked)
                mimic_loss = mimic_incorrect_pytorch_smooth_loss(
                    log_probs, log_probs_masked, target_masked, reduction, label_smoothing)
                print(f"label_smoothing = {label_smoothing}, reduction = {reduction}, "
                      f"is correct? {torch.allclose(standard_loss, masked_loss)}, "
                      f"behavior mimicked? {torch.allclose(standard_loss, mimic_loss)}")


def mimic_incorrect_pytorch_smooth_loss(log_probs, log_probs_masked, target_masked, reduction, label_smoothing):
    n_classes = log_probs.shape[-1]
    uniform_target = torch.ones_like(log_probs) / n_classes
    loss_obj = CrossEntropyLoss(reduction=reduction)
    one_hot_loss = loss_obj(log_probs_masked, target_masked)
    uniform_loss = loss_obj(log_probs, uniform_target)  # includes ignored logprobs!
    incorrect_smooth_loss = (1.0 - label_smoothing) * one_hot_loss + label_smoothing * uniform_loss
    return incorrect_smooth_loss


def _create_inputs() -> Tuple[Tensor, Tensor]:
    bsz = 100
    classes = 35
    ignore_index = -100
    num_ignore = 30

    target = torch.randint(high=classes, size=[bsz])
    ignore_mask = tensor([True] * num_ignore + [False] * (bsz - num_ignore))
    ignore_mask = ignore_mask[randperm(bsz)]
    target[ignore_mask] = ignore_index

    logits = torch.randn(bsz, classes)
    log_probs = logits.log_softmax(dim=-1)

    return log_probs, target


if __name__ == '__main__':
    label_smoothing_bug()
```

### Versions

python collect_env.py
Collecting environment information...
PyTorch version: 1.10.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-65-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: TITAN Xp
GPU 1: TITAN Xp
GPU 2: TITAN Xp
GPU 3: TITAN Xp

Nvidia driver version: 460.27.04
cuDNN version: Probably one of the following:
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.4
[pip3] pytorch-lightning==1.5.0
[pip3] torch==1.10.0+cu111
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.11.1+cu111
[conda] mypy-extensions           0.4.3                    pypi_0    pypi
[conda] numpy                     1.21.4                   pypi_0    pypi
[conda] pytorch-lightning         1.5.0                    pypi_0    pypi
[conda] torch                     1.10.0+cu111             pypi_0    pypi
[conda] torchmetrics              0.6.0                    pypi_0    pypi
[conda] torchvision               0.11.1+cu111             pypi_0    pypi

**Comment by TomerRonen34** [comment]
For now, here’s a fixed version that applies the ignore mask explicitly before calling the loss function.
```
from torch import Tensor
from torch.nn import CrossEntropyLoss


class FixedCrossEntropyLoss(CrossEntropyLoss):
    """
    Standard CrossEntropyLoss with label_smoothing doesn't handle ignore_index properly, so we apply
    the mask ourselves. See https://github.com/pytorch/pytorch/issues/73205
    """

    def forward(self, input: Tensor, target: Tensor) -> Tensor:
        if not target.is_floating_point() and self.ignore_index is not None:
            input = input[target != self.ignore_index]
            target = target[target != self.ignore_index]
        loss = super().forward(input, target)
        return loss
```

**Comment by marigoold** [comment]
Hi, I seem to have the same problem, is this bug fixed now?

**Comment by TomerRonen34** [comment]
Hi, I don't know if they fixed it yet, but you can use my FixedCrossEntropyLoss if you want to make sure that ignored samples (e.g. padding) are really ignored.

**Comment by marigoold** [comment]
Thanks for your code! BTW, I upgraded PyTorch to 1.12, and found the bug has been fixed, FYI

**Comment by jbschlosser** [comment]
Closing as fixed.
